SourceState.Apply segfaults if chezmoiexternal path is absolute

## Describe the bug

Assuming a `.chezmoiexternal.toml` entry such as:

```toml
["/home/myhome/.local.share/fonts/Hack"]
    type = "archive"
    url = "https://github.com/ryanoasis/nerd-fonts/releases/download/v2.1.0/Hack.zip"
```

`chezmoi apply` will segfault.

## Expected behavior

Two options:

* Using absolute paths in chezmoiexternal may be bad form / considered invalid user input, and `apply` could instead return a clear error.
* Or, code should be able to handle absolute paths directly.

## Output of command with the `--verbose` flag

```console
$ chezmoi --verbose apply
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x94b63f]

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/pkg/chezmoi.(*SourceState).Apply(0xc0000e9600, {0x162a010, 0xc00094e540}, {0x162a010, 0xc00094e540}, {0x16276f8, 0xc0001a0a20}, {{0xc000040155?, 0x49e746?}}, {{0x0, ...}}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/sourcestate.go:577 +0x17f
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).applyArgs(0xc0008ca000, {0x1623698, 0xc00003e0f0}, {0x162a010, 0xc00094e540}, {{0xc000040155?, 0x0?}}, {0xc0002b4dd0, 0x0, 0x1}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:634 +0x846
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).runApplyCmd(0x0?, 0x0?, {0xc0002b4dd0?, 0x0?, 0x0?})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/applycmd.go:41 +0x147
github.com/spf13/cobra.(*Command).execute(0xc000312000, {0xc0002b4dc0, 0x1, 0x1})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:872 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc00035f400)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).execute(0xc00091fdf8?, {0xc0000360a0, 0x2, 0x2})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:1201 +0x78
github.com/twpayne/chezmoi/v2/pkg/cmd.runMain({{0x1614458, 0x6}, {0x1621000, 0x28}, {0x1617c60, 0x14}, {0x1614db8, 0xa}}, {0xc0000360a0, 0x2, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:264 +0x1b4
github.com/twpayne/chezmoi/v2/pkg/cmd.Main({{0x1614458, 0x6}, {0x1621000, 0x28}, {0x1617c60, 0x14}, {0x1614db8, 0xa}}, {0xc0000360a0, 0x2, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:128 +0xb1
main.main()
	/home/runner/work/chezmoi/chezmoi/main.go:26 +0x10c
```

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                MESSAGE
ok        version              v2.21.0, commit 3af5c7fa9c744356ae5032f8a20b0aeaf13f0845, built at 2022-08-23T23:24:56Z, built by goreleaser
ok        latest-version       v2.21.0
ok        os-arch              linux/amd64 (Debian GNU/Linux)
ok        uname                Linux bistanclaque 5.18.0-4-amd64 #1 SMP PREEMPT_DYNAMIC Debian 5.18.16-1 (2022-08-10) x86_64 GNU/Linux
ok        go-version           go1.19 (gc)
ok        executable           ~/.local/bin/chezmoi
ok        upgrade-method       replace-executable
ok        config-file          ~/.config/chezmoi/chezmoi.yaml, last modified 2022-08-27T12:56:38+02:00
warning   source-dir           ~/.local/share/chezmoi is a git working tree (dirty)
ok        suspicious-entries   no suspicious entries
warning   working-tree         ~/.local/share/chezmoi is a git working tree (dirty)
ok        dest-dir             ~ is a directory
ok        shell-command        found /usr/bin/fish
ok        shell-args           /usr/bin/fish
ok        cd-command           found /usr/bin/fish
ok        cd-args              /usr/bin/fish
ok        edit-command         found /usr/bin/nvim
ok        edit-args            /usr/bin/nvim
info      diff-command         not set
ok        umask                022
ok        git-command          found /usr/bin/git, version 2.35.1
ok        merge-command        found /usr/bin/vimdiff
info      age-command          age not found in $PATH
ok        gpg-command          found /usr/bin/gpg, version 2.2.35
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keeper-command       keeper not found in $PATH
info      keepassxc-db         not set
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      vault-command        vault not found in $PATH
info      secret-command       not set
```

</details>

## Additional context

While all docs and examples mention only relative paths, I had constructed `chezmoiexternal` entries via templating (`fontsdir : {{ .chezmoi.homeDir }}/.local/share/fonts`) which masked the problem to some extent.

(I completely agree that it's simpler / more intuitive to only generate and handle relative paths in `chezmoiexternal`).

Thanks!

--- a/pkg/chezmoi/sourcestate.go
+++ b/pkg/chezmoi/sourcestate.go
@@ -16,6 +16,7 @@ import (
 	"os"
 	"os/exec"
 	"path"
+	"path/filepath"
 	"runtime"
 	"sort"
 	"strings"
@@ -1199,8 +1200,11 @@ func (s *SourceState) addExternal(sourceAbsPath AbsPath) error {
 	}
 	s.Lock()
 	defer s.Unlock()
-	for relPathStr, external := range externals {
-		targetRelPath := parentTargetSourceRelPath.JoinString(relPathStr)
+	for path, external := range externals {
+		if filepath.IsAbs(path) {
+			return fmt.Errorf("%s: %s: path is not relative", sourceAbsPath, path)
+		}
+		targetRelPath := parentTargetSourceRelPath.JoinString(path)
 		if _, ok := s.externals[targetRelPath]; ok {
 			return fmt.Errorf("%s: duplicate externals", targetRelPath)
 		}

--- a/pkg/cmd/testdata/scripts/issue2302.txt
+++ b/pkg/cmd/testdata/scripts/issue2302.txt
@@ -1,0 +1,10 @@
+[windows] skip 'UNIX only'
+
+# test that chezmoi does not panic if an external uses an absolute path
+! exec chezmoi diff
+stderr 'path is not relative'
+
+-- home/user/.local/share/chezmoi/.chezmoiexternal.toml --
+["/home/user/.dir"]
+    type = "archive"
+    url = "https://example.com/example.tar.gz"

